### DISCUSSION TITLE
Capturing the shell output in SwiftUI

### DISCUSSION THREAD

**Discussion by ghost** [migueldeicaza/SwiftTerm#308] [Q&A] [answered]
Hi,

using the LocalProcessTerminalView view I can access to the terminal to send a byte array to the shell rendered in the terminal (I use zsh).

But how can I capture the stdout of that shell inside my Swift app?

Thanks!

**Comment by migueldeicaza** [comment]
Hello,

Mhm I see that the method you would need is public, but not open, making the simplest solution hard (the `dataReceived` delegate implementation method)

An option is for me to make this open, but the other option is to manually assign the delegate on the process class and intercept all those calls and forward those to the view.

**Comment by ghost** [comment]
Hi Miguel,

thanks, I managed to capture the chars on that method. As for now i'm going to handle this manually unless you know an easier way to obtain this.

What I need to do basically is wait for some output matching a regexp before calling the send function to write something else in the terminal in response.

**Comment by migueldeicaza** [comment]
So you could override without needing a new release?

That’s perfect

**Comment by ghost** [comment]
well, I just hardcoded my things on your class. Not the most elegant and maintenable thing but in my case can work.

<img width="1324" alt="image" src="https://github.com/migueldeicaza/SwiftTerm/assets/129745430/99ba03b7-29f2-466e-8d7d-4d23755237d5">

and

<img width="1292" alt="image" src="https://github.com/migueldeicaza/SwiftTerm/assets/129745430/b98d06a4-1c1a-4b6e-93b5-f6e3a77de4cb">

but still WIP. And I still need to understand most of the decorators because I'm very new to Swift :)

**Comment by migueldeicaza** [comment]
thanks, let me add the “open” there, so you don’t need to modify my code, and instead you subclass this and modify there

**Reply by ghost** [reply]
fyi here I still needed to open the process property in that class in order to use it from the outside.

<img width="1425" alt="image" src="https://github.com/migueldeicaza/SwiftTerm/assets/129745430/c1ff25dc-370c-42c9-916e-1adab7e6b2f3">

**Comment by migueldeicaza** [comment, marked as answer]
Ok, it should be a lot easier now.

What you need to do is create a subclass of `LocalProcessTerminalView` and override the `dataReceived` method.

Like this:

```
class MyCapturingTerminalView: LocalProcessTerminalView {
    public override init (frame: CGRect)
    {
        super.init (frame: frame)
    }
    
    public required init? (coder: NSCoder)
    {
        super.init (coder: coder)
    }

    override func dataReceived(slice: ArraySlice<UInt8>) {
        super.dataReceived (slice: slice) 
        print ("Received: \(slice.count) bytes")
    }
}
```

**Reply by ghost** [reply]
Thanks. I tried to add your code in my own file, but this is what I get here:

<img width="1229" alt="image" src="https://github.com/migueldeicaza/SwiftTerm/assets/129745430/d6606f1a-660b-42b5-81a7-8fda631bbf14">

**Reply by ghost** [reply]
and also here:

<img width="1180" alt="image" src="https://github.com/migueldeicaza/SwiftTerm/assets/129745430/55fa3177-7311-4804-8e22-21ef37ecd0b6">

**Reply by migueldeicaza** [reply]
I fixed the first one, the second one I think is all on your code base

**Reply by ghost** [reply]
Yes, it was. I had to just update those two functions' types to conform the NSViewRepresentable specification.
And I saw your commits and just pulled here, everything is working as it should. Thanks
